hie-plugin-api/Haskell/Ide/Engine/ModuleCache.hs

```diff
@@ -77,75 +74,92 @@ getCradle fp = do
+withCachedModule :: FilePath -> (CachedModule -> IdeM (IdeResponse b)) -> IdeM (IdeResponse b)
+withCachedModule uri callback = do
   mcm <- getCachedModule uri
+  uri' <- liftIO $ canonicalizePath uri
   case mcm of
-    Nothing -> noCache
-    Just cm -> callback cm
+    ModuleCached cm _ -> callback cm
+    ModuleLoading -> return $ IdeResponseDeferred uri' callback
+    ModuleFailed err -> return $ IdeResponseFail err
 
 -- | Calls its argument with the CachedModule for a given URI
 -- along with any data that might be stored in the ModuleCache.
 -- The data is associated with the CachedModule and its cache is
 -- invalidated when a new CachedModule is loaded.
 -- If the data doesn't exist in the cache, new data is generated
 -- using by calling the `cacheDataProducer` function
-withCachedModuleAndData :: forall a b m.
-  (ModuleCache a, GM.MonadIO m, HasGhcModuleCache m, MonadMTState IdeState m)
-  => FilePath -> m b -> (CachedModule -> a -> m b) -> m b
-withCachedModuleAndData uri noCache callback = do
+withCachedModuleAndData :: forall a b. ModuleCache a
+                        => FilePath -> (CachedModule -> a -> IdeM (IdeResponse b)) -> IdeM (IdeResponse b)
+withCachedModuleAndData uri callback = do
   uri' <- liftIO $ canonicalizePath uri
   mcache <- getModuleCache
   let mc = (Map.lookup uri' . uriCaches) mcache
   case mc of
-    Nothing -> noCache
+    Nothing -> return $ IdeResponseDeferred uri' $ \_ -> withCachedModuleAndData uri callback
     Just UriCache{cachedModule = cm, cachedData = dat} -> do
       let proxy :: Proxy a
           proxy = Proxy
       a <- case Map.lookup (typeRep proxy) dat of
              Nothing -> do
                val <- cacheDataProducer cm
                let dat' = Map.insert (typeOf val) (toDyn val) dat
-               modifyCache (\s -> s {uriCaches = Map.insert uri' (UriCache cm dat')
+               modifyCache (\s -> s {uriCaches = Map.insert uri' (UriCache cm dat' False)
                                                                  (uriCaches s)})
                return val
-             Just x -> do
+             Just x ->
                case fromDynamic x of
                  Just val -> return val
                  Nothing  -> error "impossible"
       callback cm a
 
 -- | Saves a module to the cache
-cacheModule :: (GM.MonadIO m, HasGhcModuleCache m)
-            => FilePath -> CachedModule -> m ()
+cacheModule :: FilePath -> CachedModule -> IdeGhcM ()
 cacheModule uri cm = do
   uri' <- liftIO $ canonicalizePath uri
+
   modifyCache (\gmc ->
       gmc { uriCaches = Map.insert
                           uri'
-                          (UriCache cm Map.empty)
+                          (UriCache cm Map.empty False)
                           (uriCaches gmc)
           }
     )
-  where
+
+  -- execute any queued actions for the module
+  actions <- fmap (fromMaybe [] . Map.lookup uri') (requestQueue <$> readMTS)
+  liftToGhc $ forM_ actions (\(_, a) -> a cm)
```

Comment:
Shouldn't we be checking if the action has been cancelled before executing it?

Reply:
I believe this is checked in `Dispatcher.hs` https://github.com/Bubba/haskell-ide-engine/blob/edc5abe06dde60eee11a23b7008d17288ec4bd4f/src/Haskell/Ide/Engine/Dispatcher.hs#L80

Reply:
Then we don't need to have the LspId in the deferred request queue, right?

Reply:
Would seem so! Removing now